cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -0,0 +1,284 @@
+};
+
+/**
+ * @brief Maps an `aggregation::Kind` value to it's corresponding binary
+ * operator.
+ *
+ * @note Not all values of `aggregation::Kind` have a valid corresponding binary
+ * operator. For these values `E`,
+ * `std::is_same_v<corresponding_operator<E>::type, void>`.
+ *
+ * @tparam k The `aggregation::Kind` value to map to its corresponding operator
+ */
+template <aggregation::Kind k>
+struct corresponding_operator {
+  using type = void;
+};
+template <>
+struct corresponding_operator<aggregation::MIN> {
+  using type = DeviceMin;
+};
+template <>
+struct corresponding_operator<aggregation::MAX> {
+  using type = DeviceMax;
+};
+template <>
+struct corresponding_operator<aggregation::SUM> {
+  using type = DeviceSum;
+};
+
+template <aggregation::Kind k>
+using corresponding_operator_t = typename corresponding_operator<k>::type;
+
+/**---------------------------------------------------------------------------*
+ * @brief Determines accumulator type based on input type and aggregation.
+ *
+ * @tparam Source The type on which the aggregation is computed
+ * @tparam k The aggregation performed
+ *---------------------------------------------------------------------------**/
+template <typename Source, aggregation::Kind k, typename Enable = void>
+struct target_type_impl {
+  using type = void;
+};
+
+// Computing MIN of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MIN> {
+  using type = Source;
+};
+
+// Computing MAX of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MAX> {
+  using type = Source;
+};
+
+// Always use size_type accumulator for COUNT
+template <typename Source>
+struct target_type_impl<Source, aggregation::COUNT> {
+  using type = cudf::size_type;
+};
+
+// Always use `double` for MEAN
+template <typename Source>
+struct target_type_impl<Source, aggregation::MEAN> {
+  using type = double;
+};
+
+// Summing integers of any type, always use int64_t accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::SUM,
+                        std::enable_if_t<std::is_integral<Source>::value>> {
+  using type = int64_t;
+};
+
+// Summing float/doubles, use same type accumulator
+template <typename Source>
+struct target_type_impl<
+    Source, aggregation::SUM,
+    std::enable_if_t<std::is_floating_point<Source>::value>> {
+  using type = Source;
+};
+
+// Always use `double` for quantile
+template <typename Source>
+struct target_type_impl<Source, aggregation::QUANTILE> {
+  using type = double;
+};
+
+// MEDIAN is a special case of a QUANTILE
+template <typename Source>
+struct target_type_impl<Source, aggregation::MEDIAN> {
+  using type = target_type_impl<Source, aggregation::QUANTILE>;
+};
+
+// Always use `size_type` for ARGMAX index
+template <typename Source>
+struct target_type_impl<Source, aggregation::ARGMAX> {
```

Comment:
Will it be helpful if we add ARGMAX and ARGMIN to handle non-fixed and fixed-width differently, as per my understanding, the current method would require a gather later once we have gather map. But, this would be helpful in case of `string_view`, but in case of fixed_width types, this would be an additional over head.

Reply:
Fixed vs. non-fixed has no bearing on the resulting type of an argmin/max aggregation. The result is always a row-index, which means it is always int32.

The `gather` step is only important for computing a min/max via an argmin/max, which is handled independently.

Reply:
Ahh, I got confused.